mesonpy/__init__.py

```diff
@@ -995,6 +996,34 @@ def _project(config_settings: Optional[Dict[Any, Any]]) -> Iterator[Project]:
         yield project
 
 
+def _get_ninja_if_needed() -> List[str]:
+    env_ninja = os.environ.get('NINJA', None)
+    ninja_candidates = [env_ninja] if env_ninja else ['ninja', 'ninja-build', 'samu']
+    for ninja in ninja_candidates:
+        ninja_path = shutil.which(ninja)
+        if ninja_path is None:
+            continue
+
+        result = subprocess.run([ninja_path, '--version'], check=False, text=True, capture_output=True)
+
+        try:
+            candidate_version = tuple(int(x) for x in result.stdout.split('.')[:3])
+        except ValueError:
+            # The meson search function skips forward if the version is not readable or too low
```

Comment:
I don't get the reference to meson in this comment. I would simply drop the comment.

Reply:
This algorithm was copied from meson. Maybe linking to the lines would be better?

Reply:
I think the algorithm is straightforward and does not need explanation. A link to the implementation in meson is doomed to become stale quickly.

Reply:
It can be a permalink. The idea though is this ideally needs to match, since it needs to evaluate the version of the same ninja that meson would. It doesn't have to be perfect, but it's best if it's as close to in-sync as possible.

Reply:
The permalink has the drawback that it will document only the algorithm as implemented in the current version of Meson and will not take into account possible future changes to it. If Meson ever changes, the link would actually be detrimental. I would just document the requirement at the beginning of the function. Something like:

```python
# Verify that a suitable ``ninja`` command is available. This function needs and 
# the corresponding one in Meson need to resolve to the same executable.
# Mimic as close as possible the algorithm used by Meson.
```

However, this brings another question: why do we need to delegate to Meson again running `ninja`? If we have this function that finds a suitable `ninja` command, we can simply run `ninja -C ...` ourselves instead than going through meson to do the same for us. Or does `meson setup` fail if `ninja` is not available?

Reply:
I believe meson setup looks for ninja, so if it was off enough, it could be problematic. I wonder if meson could expose the ninja search for meson-python to just use?

Reply:
Meson does not expose a public documented python API, this is an intentional decision.

But it's honestly not all that important. The Meson algorithm is just about covering some common bases to ensure that some sort of ninja is found. meson-python wants to support samurai because it's right for meson-python to allow it -- not because it "has to since Meson does".

Reply:
@eli-schwartz 's and mine comments crossed each other...

Reply:
@eli-schwartz This function only check that a `ninja` like command exist. It install `ninja` form PyPI if it does not. It leaves it to meson to invoke it. Thus if this function covers something that Meson does not, we have a problem.

Reply:
- Meson might add support for things that this then misses out on, but the worst outcome is installing an extra ninja via pip, so nothing will break
- meson-python can guarantee that if it found a good ninja, Meson always find it too -- simply export $NINJA pointing to the executable you want. Even though meson will of course find it anyway because it engages in the same general approach to sensible checking... You can guarantee it beyond all shadow of a doubt.


Meson's python API may be internal and subject to change if someone decides that for refactoring purposes it would look better somewhere else...

... but the $NINJA environment variable is a public user-facing feature and can absolutely 100% be relied on.

Reply:
So instead of returning true/false, this can return the location of the discovered ninja, then $NINJA can be set with it. Still is truthy based on if it was found.